tfx_bsl/beam/run_inference.py

```diff
@@ -221,17 +232,207 @@ def update(self, elements: List[Union[tf.train.Example,
+@beam.typehints.with_output_types(prediction_log_pb2.PredictionLog)
+class _RemotePredictDoFn(_BaseDoFn):
+  """A DoFn that performs predictions from a cloud-hosted TensorFlow model.
+
+  Supports both batch and streaming processing modes.
+  NOTE: Does not work on DirectRunner for streaming jobs [BEAM-7885].
+
+  In order to request predictions, you must deploy your trained model to AI
+  Platform Prediction in the TensorFlow SavedModel format. See
+  [Exporting a SavedModel for prediction]
+  (https://cloud.google.com/ai-platform/prediction/docs/exporting-savedmodel-for-prediction)
+  for more details.
+  """
+  def __init__(self, inference_endpoint: model_spec_pb2.InferenceEndpoint):
+    super(_RemotePredictDoFn, self).__init__()
+    self._api_client = None
+    project_id = inference_endpoint.model_endpoint_spec.project_id
+    if not project_id:
+      raise ValueError('A non-empty project id must be provided.')
+
+    model_name = inference_endpoint.model_endpoint_spec.model_name
+    if not model_name:
+      raise ValueError('A non-empty model name must be provided.')
+
+    version_name = inference_endpoint.model_endpoint_spec.version_name
+    name_spec = 'projects/{}/models/{}'
+    # If version is not specified, the default version for a model is used.
+    if version_name:
+      name_spec += '/versions/{}'
+    self.full_model_name = name_spec.format(project_id, model_name,
+                                            version_name)
+
+  def setup(self):
+    super(_RemotePredictDoFn, self).setup()
+    self._api_client = self._get_api_client()
+
+  @staticmethod
+  def _get_api_client() -> discovery.Resource:
+    return discovery.build('ml', 'v1')
+
+  @staticmethod
+  def _execute_request(request: http.HttpRequest) -> dict:
+    return request.execute()
+
+  def _make_request(self, model_name: str, body: dict) -> http.HttpRequest:
+    return self._api_client.projects().predict(name=model_name, body=body)
+
+  @classmethod
+  def _prepare_instances(cls,
+    elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> Generator[Mapping[Text, Any], None, None]:
+    for example in elements:
+      # TODO: support tf.train.SequenceExample
+      if not isinstance(example, tf.train.Example):
+        raise ValueError('Remote prediction only supports tf.train.Example')
+
+      instance = {}
+      for input_name, feature in example.features.feature.items():
+        attr_name = feature.WhichOneof('kind')
+        attr = getattr(feature, attr_name)
+        values = cls._parse_feature_content(attr.value, attr_name,
+                                            cls._sending_as_binary(input_name))
+        values = cls._maybe_flatten(values)
+        instance[input_name] = values
+      yield instance
+
+  @staticmethod
+  def _sending_as_binary(input_name: Text) -> bool:
+    """Whether data should be sent as binary."""
+    return input_name.endswith('_bytes')
```

Comment:
Could you add the description of "_bytes" suffix to the _RemotePredictDoFn comment? Thanks

Reply:
Done. Thanks!